File: tests/TestBindings/Lua/TestUpgradeType.lua
function constructors()
  unknown = BWAPI.UpgradeType(BWAPI.UpgradeTypes.Enum.Unknown)
  weapons = BWAPI.UpgradeType(BWAPI.UpgradeTypes.Enum.Terran_Infantry_Weapons)
end

function equality()
  unknown = BWAPI.UpgradeType(BWAPI.UpgradeTypes.Enum.Unknown)
  weapons = BWAPI.UpgradeType(BWAPI.UpgradeTypes.Enum.Terran_Infantry_Weapons)

  assert(unknown ~= weapons)
  assert(unknown == BWAPI.UpgradeTypes.Unknown)
  assert(weapons == BWAPI.UpgradeTypes.Terran_Infantry_Weapons)
end

function toStringTest()
  unknown = BWAPI.UpgradeType(BWAPI.UpgradeTypes.Enum.Unknown)
  weapons = BWAPI.UpgradeType(BWAPI.UpgradeTypes.Enum.Terran_Infantry_Weapons)

  assert(tostring(unknown) == 'Unknown')
  assert(tostring(weapons) == 'Terran_Infantry_Weapons')
end

function valuesOfTheWeaponsUpgradeType()
  upgrade = BWAPI.UpgradeTypes.Terran_Infantry_Weapons

  assert(upgrade:getID() == BWAPI.UpgradeTypes.Enum.Terran_Infantry_Weapons)
  assert(upgrade:getName() == "Terran_Infantry_Weapons")
  assert(upgrade:getRace() == BWAPI.Races.Terran)
  assert(upgrade:mineralPrice() == 100)
  assert(upgrade:mineralPrice() + upgrade:mineralPriceFactor() == upgrade:mineralPrice(2))
  assert(upgrade:gasPrice() == 100)
  assert(upgrade:gasPrice() + upgrade:gasPriceFactor() == upgrade:gasPrice(2))
  assert(upgrade:upgradeTime() == 4000)
  assert(upgrade:upgradeTime() + upgrade:upgradeTimeFactor() == upgrade:upgradeTime(2))
  assert(upgrade:maxRepeats() == 3)
  assert(upgrade:whatUpgrades() == BWAPI.UnitTypes.Terran_Engineering_Bay)
  assert(upgrade:whatsRequired() == BWAPI.UnitTypes.None)
  assert(upgrade:whatsRequired(1) == BWAPI.UnitTypes.None)
  assert(upgrade:whatsRequired(2) == BWAPI.UnitTypes.Terran_Science_Facility)
  assert(#upgrade:whatUses() == 11)
  assert(upgrade:whatUses():contains(BWAPI.UnitTypes.Terran_Marine))
end

function valuesOfTheUnknownUpgradeType()
  upgrade = BWAPI.UpgradeTypes.Unknown

  assert(upgrade:getID() == BWAPI.UpgradeTypes.Enum.Unknown)
  assert(upgrade:getName() == "Unknown")
  assert(upgrade:getRace() == BWAPI.Races.Unknown)
  assert(upgrade:mineralPrice() == 0)
  assert(upgrade:mineralPrice() + upgrade:mineralPriceFactor() == upgrade:mineralPrice(2))
  assert(upgrade:gasPrice() == 0)
  assert(upgrade:gasPrice() + upgrade:gasPriceFactor() == upgrade:gasPrice(2))
  assert(upgrade:upgradeTime() == 0)
  assert(upgrade:upgradeTime() + upgrade:upgradeTimeFactor() == upgrade:upgradeTime(2))
  assert(upgrade:maxRepeats() == 0)
  assert(upgrade:whatUpgrades() == BWAPI.UnitTypes.None)
  assert(upgrade:whatsRequired() == BWAPI.UnitTypes.None)
  assert(upgrade:whatsRequired(1) == BWAPI.UnitTypes.None)
  assert(upgrade:whatsRequired(2) == BWAPI.UnitTypes.None)
  assert(#upgrade:whatUses() == 0)
end

File: tests/TestBindings/TestUpgradeType.cpp
#include <catch.hpp>
#include <BWAPI/UnitType.h>
#include <sol.hpp>
#include <BWAPI_Lua.h>
#include "BindingTestHelpers.h"

using namespace BWAPI;
using namespace BWAPI_Lua;

TEST_CASE("upgrade type bindings", "[upgradetype]")
{
	sol::state lua;
	lua.open_libraries(sol::lib::base);
	bindUpgradeType(lua.create_named_table("BWAPI"));
	lua.script_file(TestBindings::getLuaTestPath(__FILE__));

	SECTION("constructors")
	{
		sol::function test = lua["constructors"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("equality")
	{
		sol::function test = lua["equality"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("tostring")
	{
		sol::function test = lua["toStringTest"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("upgrade type values")
	{
		bindRace(lua["BWAPI"]);
		bindTechType(lua["BWAPI"]);
		bindUpgradeType(lua["BWAPI"]);
		bindUnitSizeType(lua["BWAPI"]);
		bindPosition(lua["BWAPI"]);
		bindWeaponType(lua["BWAPI"]);
		bindUnitType(lua["BWAPI"]);

		SECTION("weapons")
		{
			sol::function test = lua["valuesOfTheWeaponsUpgradeType"];
			REQUIRE_NOTHROW(test());
		}

		SECTION("unknown")
		{
			sol::function test = lua["valuesOfTheUnknownUpgradeType"];
			REQUIRE_NOTHROW(test());
		}
	}
}